c10 dispatcher stores autograd kernels

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23695 [wip] Fixed size arrays
* #23668 [wip] Call aten ops through c10 dispatcher
* #23888 Allow TensorMethods.h to include Dispatcher.h (alternative)
* #23667 Register ATen ops with c10
* #23748 Disambiguate tensor and string ops
* #23694 toString(FunctionSchema) shows overload name
* **#23666 c10 dispatcher stores autograd kernels**
* #23665 Allow kernels that don't have a boxed version
* #23447 Unboxed kernels in c10
* #23834 Open up AliasAnalysisKind for any ops

Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
[wip] c10 dispatcher stores autograd kernels



Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "[wip] c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

dzhulgakov commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -87,13 +94,30 @@ class OperatorEntry final {
+  // towards a system where ops register autograd wrappers (i.e. functions that
+  // do some wrapping code and get a pointer to the actual kernel) instead of
+  // autograd functions.
+  // This is a list because, similar to kernels_, multiple libraries could
```
why this complexity? if this is a temp solution - just keep it simple and have a single pointer. For multiple wrappers we'd need to build something different. But even in that case I think errorring out instead of creating a list is the right behavior

smessmer replied:
it mirrors what we're doing for kernels, i.e. allow overwriting already existing kernels but in a safe way and reset them once the overwriting library is gone. A potential use for this would be XLA overwriting the default autograd kernels provided by ATen.

I think the complexity is worth it - even though this is hopefully a temporary solution, these kind of solutions tend to stick around and cause issues if not done right.

dzhulgakov replied:
as discussed - let's extend this comment to note that we primarily want to enable iPython notebooks. Otherwise, this behavior is actually bad - as you want to get errors, not silent overwrite on collision.

smessmer replied:
I added a comment

dzhulgakov commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -87,13 +94,30 @@ class OperatorEntry final {
+  // before older registrations and the list head is the autograd kernel
+  // which is currently used.
+  std::list<void*> unboxedAutogradKernels_;
+  std::mutex unboxedAutogradKernelsMutex_;
```
again, I think it's fine to assume that library loads don't happen concurrently. Even if they did - having a top-level mutex is a much easier solution that also doesn't blow up OperatorEntry size.

btw, since you already have kernelsMutex_ above - why not just use that (even though we don't really need mutex)

smessmer replied:
The mutex costs us even less, the complexity added by just the mutex it is negligible, but then we get parallel and race-free library loading for free, don't see a reason for not doing it.

The default pattern for mutexes is to bind them directly to the data they protect and trying to keep these data units small. OperatorEntry size doesn't matter, this is not an entry in the dispatch table.

dzhulgakov reviewed (changes requested):
It seems a bit over-engineering for the purpose and it makes code much less understandable.

smessmer committed:
Update on "c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

dzhulgakov commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -87,13 +94,30 @@ class OperatorEntry final {
   > kernels_;
   std::mutex kernelsMutex_; // protects kernels_
```
just to be paranoid - move mutexes towards the end, so that read path hits the same cache line. Though we still going to do another jump for the std::list or read from the atomic :(

smessmer committed:
Update on "c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

dzhulgakov reviewed (approved):
Thanks for the comments!

smessmer committed:
Update on "c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)

smessmer committed:
Update on "c10 dispatcher stores autograd kernels"


Differential Revision: [D16603130](https://our.internmc.facebook.com/intern/diff/D16603130/)
